Removing Hardlink from Mode property in default file system format

powercode opened the pull request:
Adding a new table view, `childrenWithHardlink`, that has the same behavior as `children` view had before this PR.

The `children` view now uses the modified `ModeWithoutHardlink` property that doesn't check for hardlinks when setting the `l` letter of `Mode` column. 

The `Mode` property keeps it's existing behavior, so this change should not be breaking.



## PR Summary

https://github.com/PowerShell/PowerShell/issues/8051

Significant speedup for default filesystem output
```PowerShell
$files = gci -Recurse c:\windows -ea:0
$files | ft | out-null
$files | ft -view childrenWithHardlink  | out-null
```
view | time | speedup
-- | -- | --
childrenWithHardlink | 3m22s | 1x
 children | 7.6s | 26x

## Formatting result

### Output before
```output
PS> gci

    Directory: D:\repos\PowerShell\symlinks


Mode                LastWriteTime         Length Name
----                -------------         ------ ----
d-----       2019-02-03     15:37                dir
d----l       2019-02-03     15:38                symdir
d----l       2019-02-03     15:39                symsubdir
-a----       2019-02-03     15:38              3 file
-a---l       2019-02-03     15:37              3 hardlink
-a---l       2019-02-03     15:38              0 symfile1
```
### Output after
```output
PS> gci

    Directory: D:\repos\PowerShell\symlinks

Mode                LastWriteTime         Length Name
----                -------------         ------ ----
d----         2019-02-03    15:37                dir
l----         2019-02-03    15:38                symdir -> D:\repos\PowerShell\symlinks\dir
l----         2019-02-03    15:39                symsubdir -> D:\repos\PowerShell\symlinks\dir\subdir\
-a---         2019-02-03    15:38              3 file
-a---         2019-02-03    15:37              3 hardlink
la---         2019-02-03    15:38              0 symfile1 -> D:\repos\PowerShell\file

PS> gci | ft -view childrenWithHardlink

    Directory: D:\repos\PowerShell\symlinks

Mode                LastWriteTime         Length Name
----                -------------         ------ ----
d----         2019-02-03    15:37                dir
l----         2019-02-03    15:38                symdir -> D:\repos\PowerShell\symlinks\dir
l----         2019-02-03    15:39                symsubdir -> D:\repos\PowerShell\symlinks\dir\subdir\
-a---         2019-02-03    15:38              3 file
la---         2019-02-03    15:37              3 hardlink
la---         2019-02-03    15:38              0 symfile1 -> D:\repos\PowerShell\file
```
produces the old output.
## PR Context  

<!-- Provide a little reasoning as to why this Pull Request helps and why you have opened it. -->

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.  
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**  
    - [x] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [x] Issue filed: https://github.com/MicrosoftDocs/PowerShell-Docs/issues/3665
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [ ] [Add `[feature]` to your commit messages if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

powercode committed:
Removing Hardlink from Mode property in default file system format

powercode committed:
Fixing $format.FormatViewDefinition | Should -HaveCount 4.

Since a new format view was added.

SteveL-MSFT commented on src/System.Management.Automation/FormatAndOutput/DefaultFormatters/FileSystem_format_ps1xml.cs:
```diff
@@ -67,6 +68,23 @@ private static IEnumerable<FormatViewDefinition> ViewsOf_FileSystemTypes(CustomC
+                    .StartRowDefinition(wrap: true)
+                        .AddPropertyColumn("Mode")
+                        .AddScriptBlockColumn(@"
+                                        [String]::Format(""{0,10} {1,8}"", $_.LastWriteTime.ToString(""d""), $_.LastWriteTime.ToString(""t""))
```
Seems like this is indented more than necessary

powercode replied:
I think I was just following the formatting/style of the rest of the file. Not sure what we want here.

powercode replied:
Oh, you are right! fixing.

powercode committed:
Fixing Typeinference test

SteveL-MSFT commented on src/System.Management.Automation/namespaces/FileSystemProvider.cs:
```diff
@@ -1864,7 +1879,7 @@ public static string Mode(PSObject instance)
             bool isReparsePoint = InternalSymbolicLinkLinkCodeMethods.IsReparsePoint(fileInfo);
-            bool isHardLink = InternalSymbolicLinkLinkCodeMethods.IsHardLink(fileInfo);
+            bool isHardLink = includeHardLink ? InternalSymbolicLinkLinkCodeMethods.IsHardLink(fileInfo) : false;
             mode[5] = isReparsePoint || isHardLink ? 'l' : '-';
```
Might be better to not include the `-` if `ModeWithoutHardlink` is used to not imply it's not a hardlink.

powercode replied:
True. Will fix

powercode replied:
@SteveL-MSFT  The result may then look like this:

```output
Mode                LastWriteTime         Length Name
----                -------------         ------ ----
d----         2019-01-29    15:29                .vsts-ci
d----l        2019-01-30    17:44                a
d----         2018-06-18    23:11                demos
```
`a` is a symlink, but since we don't show the '-' it will be hanging.

SteveL-MSFT replied:
Let's have @PowerShell/powershell-committee review this and propose something

powercode replied:
@SteveL-MSFT And when it is a file? That is hardlinked?

This is how they look today.
```output
-a---l        2018-09-15    18:29         316640 WMSysPr9.prx
-a---l        2018-09-15    09:29          11264 write.exe
```
Should that be like this?
```output
la---        2018-09-15    18:29         316640 WMSysPr9.prx
la---        2018-09-15    09:29          11264 write.exe
```

SteveL-MSFT replied:
@powercode yes (`la---`), the intent is to reuse the 'directory' column since a link is a different type that a file or directory.

SteveL-MSFT replied:
I believe this is resolved.  Following the Unix `ls -l` model, a link is distinct from a file or directory.  You don't know what it points to until you resolve it.  `ls -l` solves this with the `link -> target` format which you've added.

powercode committed:
Adressing Steve's feedback

adityapatwardhan commented on src/System.Management.Automation/FormatAndOutput/DefaultFormatters/FileSystem_format_ps1xml.cs:
```diff
@@ -2,6 +2,7 @@
 // Licensed under the MIT License.
 
 using System.Collections.Generic;
+using Microsoft.PowerShell.Commands;
```
Is this required?

adityapatwardhan commented on src/System.Management.Automation/namespaces/FileSystemProvider.cs:
```diff
@@ -1847,6 +1847,21 @@ private FlagsExpression<FileAttributes> FormatAttributeSwitchParameters()
+        [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Reliability", "CA2001:AvoidCallingProblematicMethods")]
+        public static string ModeWithoutHardlink(PSObject instance)
+        {
+            return Mode(instance, includeHardLink: false);
```
The summary comment says `Provides a mode property for FileSystemInfo including HardLinks` but the code says `includeHardLink: false`, this is very confusing.

powercode replied:
Copy and paste - fixing.

adityapatwardhan commented:
@powercode The change looks good in general. Could you also put a before and after of how the output of `Get-ChildItem` looks like for future reference in the PR description.

adityapatwardhan commented:
@powercode Could also have a look at the CodeFactor issue, some on them are about documentation and formating. Would be nice to get those fixed.

powercode committed:
Addressing Aditya's feedback.

Also fixing whitespace issues.

JamesWTruher commented:
symlinks will still appear in the default view, right?
j

powercode commented:
@JamesWTruher Yes, symlinks appear as a `l`.

adityapatwardhan commented:
@powercode Restarted the CI.

adityapatwardhan commented:
@powercode Please respond to the committee recommendation about re-using the `d` field for links instead of adding one at the end.

SteveL-MSFT commented:
@powercode no, that's perfect!

powercode commented:
@SteveL-MSFT I'm playing now with moving AddScriptBlockColumn into hidden code properties. Seems to be a bit faster to.

Any reason not to do that? The Name string needed it to get to the internal methods for Link management.

powercode committed:
Using dir column for link info.

powercode committed:
Adding display of symlink target.
CodeProperties for file system formatting for speed.

powercode committed:
Slightly faster LastWriteTime - noticable in profiler!

powercode committed:
CodeFactor fixes

powercode commented:
A bit of feature creep here :) Feel free to pick the commits you want!

adityapatwardhan commented:
@powercode Please have a look at the test failure

SteveL-MSFT reviewed (changes requested):
We should have some tests validating the formating changes

powercode committed:
NameString -> TargetName

Reordering mode setter lines.
Fixing test

powercode committed:
Fixing xUnit Mode test

SteveL-MSFT commented on src/System.Management.Automation/namespaces/FileSystemProvider.cs:
```diff
@@ -1864,12 +1879,63 @@ public static string Mode(PSObject instance)
+        /// </summary>
+        /// <param name="instance">Instance of PSObject wrapping a FileSystemInfo.</param>
+        /// <returns>Name if a file or directory, Name -> Target if symlink.</returns>
+        public static string NameString(PSObject instance)
```
Got it

powercode committed:
CodeFactor whitespace

powercode commented on src/System.Management.Automation/namespaces/FileSystemProvider.cs:
```diff
@@ -1864,12 +1879,63 @@ public static string Mode(PSObject instance)
-            bool isHardLink = InternalSymbolicLinkLinkCodeMethods.IsHardLink(fileInfo);
-            mode[5] = isReparsePoint || isHardLink ? 'l' : '-';
+            bool isHardLink = includeHardLink ? InternalSymbolicLinkLinkCodeMethods.IsHardLink(fileInfo) : false;
+            mode[0] = isReparsePoint || isHardLink ? 'l' : mode[0];
```
Fixed

powercode commented on src/System.Management.Automation/namespaces/FileSystemProvider.cs:
```diff
@@ -1864,12 +1879,63 @@ public static string Mode(PSObject instance)
             // Mark the last bit as a "l" if it's a reparsepoint (symbolic link or junction)
```
removed

powercode commented:
I consider the remaining CodeFactor issues to be noise, and the CI static analysis seem to complain about spelling in MD files.

SteveL-MSFT commented on src/System.Management.Automation/namespaces/FileSystemProvider.cs:
```diff
@@ -1845,31 +1845,97 @@ private FlagsExpression<FileAttributes> FormatAttributeSwitchParameters()
+        /// </summary>
+        /// <param name="instance">Instance of PSObject wrapping a FileSystemInfo.</param>
+        /// <returns>Name if a file or directory, Name -> Target if symlink.</returns>
+        public static string TargetName(PSObject instance)
```
Sorry, I misunderstood what this was used for and thought it was only used to render target for links.  This should be renamed back to `NameString`

iSazonov reviewed (commented):
@SteveL-MSFT Has PowerShell Committee approved new public APIs?

powercode committed:
TargetName -> NameString

powercode commented on src/System.Management.Automation/namespaces/FileSystemProvider.cs:
```diff
@@ -1845,31 +1845,97 @@ private FlagsExpression<FileAttributes> FormatAttributeSwitchParameters()
-            bool isHardLink = InternalSymbolicLinkLinkCodeMethods.IsHardLink(fileInfo);
-            mode[5] = isReparsePoint || isHardLink ? 'l' : '-';
 
             return new string(mode);
```
That's an assignment, no allocation. But I'm rewriting that code to be a bit more clear.

powercode committed:
Addressing Ilya's feedback

powercode commented:
This is what the default view and the `childrenWithHardlink`  view looks like:

```output
PS> gci
    Directory: D:\repos\PowerShell\symlinks

Mode                LastWriteTime         Length Name
----                -------------         ------ ----
d----         2019-02-03    15:37                dir
l----         2019-02-03    15:38                symdir -> D:\repos\PowerShell\symlinks\dir
l----         2019-02-03    15:39                symsubdir -> D:\repos\PowerShell\symlinks\dir\subdir\
-a---         2019-02-03    15:38              3 file
-a---         2019-02-03    15:37              3 hardlink
la---         2019-02-03    15:38              0 symfile1 -> D:\repos\PowerShell\file
```
```output
PS> gci | ft -view childrenWithHardlink

    Directory: D:\repos\PowerShell\symlinks

Mode                LastWriteTime         Length Name
----                -------------         ------ ----
d----         2019-02-03    15:37                dir
l----         2019-02-03    15:38                symdir -> D:\repos\PowerShell\symlinks\dir
l----         2019-02-03    15:39                symsubdir -> D:\repos\PowerShell\symlinks\dir\subdir\
-a---         2019-02-03    15:38              3 file
la---         2019-02-03    15:37              3 hardlink
la---         2019-02-03    15:38              0 symfile1 -> D:\repos\PowerShell\file
```

powercode committed:
Updating expected string when TargetName changed to NameString.

iSazonov commented on src/System.Management.Automation/namespaces/FileSystemProvider.cs:
```diff
@@ -1845,29 +1845,81 @@ private FlagsExpression<FileAttributes> FormatAttributeSwitchParameters()
+        {
+            return instance?.BaseObject is FileSystemInfo fileInfo
+                ? InternalSymbolicLinkLinkCodeMethods.IsReparsePoint(fileInfo)
+                    ? $"{fileInfo.Name} -> {InternalSymbolicLinkLinkCodeMethods.GetTarget(instance).First()}"
```
Please use explicit formatting with explicit Culture.
We have an Issue to review using Culture in all code base and this help us.

iSazonov commented on src/System.Management.Automation/namespaces/FileSystemProvider.cs:
```diff
@@ -1845,29 +1845,81 @@ private FlagsExpression<FileAttributes> FormatAttributeSwitchParameters()
+        {
+            return instance?.BaseObject is FileInfo fileInfo
+                ? fileInfo.Attributes.HasFlag(FileAttributes.Offline)
+                    ? $"({fileInfo.Length})"
```
Please use explicit formatting with explicit Culture.
We have an Issue to review using Culture in all code base and this help us.

powercode replied:
It's just an integer. When would culture matter with an integer?

powercode replied:
Besides, this is for UI formatting, so we want currentculture anyway, don't we?

iSazonov replied:
There may be exceptions.
In any case, it would be more convenient for us to have one pattern - we are thinking about conducting an audit of using Cultures and this would facilitate our work.

iSazonov commented on src/System.Management.Automation/namespaces/FileSystemProvider.cs:
```diff
@@ -1845,29 +1845,81 @@ private FlagsExpression<FileAttributes> FormatAttributeSwitchParameters()
+                    fileAttributes.HasFlag(FileAttributes.Hidden)    ? 'h' : '-',
+                    fileAttributes.HasFlag(FileAttributes.System)    ? 's' : '-',
+                };
+                return new string(mode);
```
I still don't understant why we need "return new"
```suggestion
                return mode;
```

powercode replied:
So that we return a string, and not a char[].

iSazonov replied:
Oh, I'm stupid :-) 
We have great new [String.Create()](https://docs.microsoft.com/en-us/dotnet/api/system.string.create?view=netcore-2.2) for the case.

powercode replied:
You can see the implementation here: https://github.com/dotnet/corefx/blob/a948042501969cf1968bec5fec998435c10e1bae/src/Common/src/CoreLib/System/String.cs#L332

Both do the same thing, but the ctor I'm using is doing it with less overhead.

powercode replied:
The temporary allocation on the stack, (the stackalloc) is just a part of the frame setup. You can consider it free. If you can show me that the string create version is faster, I'll be happy to change it, but I would be very surprised if that is the case.

iSazonov replied:
:-) Ah, yes, I spent a huge time on perf benchmark tests while I worked on Simple Case Folding implementation that it became a habit :-)

``` ini

BenchmarkDotNet=v0.11.3, OS=Windows 10.0.17763.253 (1809/October2018Update/Redstone5)
Intel Core i3-4150 CPU 3.50GHz (Haswell), 1 CPU, 4 logical and 2 physical cores
.NET Core SDK=2.1.403
  [Host]     : .NET Core 2.1.5 (CoreCLR 4.6.26919.02, CoreFX 4.6.26919.02), 64bit RyuJIT
  DefaultJob : .NET Core 2.1.5 (CoreCLR 4.6.26919.02, CoreFX 4.6.26919.02), 64bit RyuJIT


```
|       Method |     Mean |     Error |    StdDev | Ratio | RatioSD |
|------------- |---------:|----------:|----------:|------:|--------:|
| SpanToString | 18.92 ns | 0.2005 ns | 0.1675 ns |  1.00 |    0.00 |
| StringCreate | 11.82 ns | 0.3090 ns | 0.3306 ns |  0.63 |    0.02 |

String.Create() is very fast.
I expected it (why would Core team need to blow this method anyway?) but it's always better to check.

iSazonov replied:
> The temporary allocation on the stack, (the stackalloc) is just a part of the frame setup.

I'm not sure if this works for the used pattern with initiator.
For `ReadOnlySpan<char> mode = stackalloc char[4]` (use explicit constant) it works exactly.

powercode replied:
After looking into this quickly, it seems like string.Create might be slightly faster, but at the expense of creating a delegate. 
The delegate can be precreated and stored static, but then the clarity of the code is worse. 
The perf improvment is about 1e-8 seconds, which in my opinion is not worth it.

iSazonov replied:
My metering shows 38% win - Is this the purpose of this PR to make this method faster? :-)

powercode committed:
stackallocing Mode array to reduce allocations

powercode committed:
Tests: Fixing casing in HardLink.

powercode committed:
ReadonlySpan instead of Span in Mode alloc.

adityapatwardhan commented:
@SteveL-MSFT Can you update your review?
@iSazonov are your comments blocking?

iSazonov commented:
@adityapatwardhan My metering shows 38% perf win in the local method. But we can not block and merge "as is" then I would push new PR.

powercode committed:
Reducing allocations in Mode by returning constants for the common cases.

powercode commented:
I pushed another commit that removes the discussed string allocations in 95+% of the cases. At the expense of clarity ;)

powercode commented:
The CI failure on windows seems like  infrastructure related - not caused by me.

iSazonov commented:
@powercode Thanks for great contribution! 
I restarted CI-Windows.

SteveL-MSFT reviewed (changes requested):
Can we have a few tests to validate the formatting so there isn't an accidental regression in the future?

powercode commented:
Sure

powercode committed:
fixing interop-definition for CreateSymbolicLink.

using bool instead of int as the return type.

The definition is:
BOOLEAN CreateSymbolicLinkA(
  LPCSTR lpSymlinkFileName,
  LPCSTR lpTargetFileName,
  DWORD  dwFlags
);

And BOOLEAN is a typedef for BYTE.

So marshalling it as an int may lead to trouble.

powercode committed:
Adding tests for FileSsytemProvider.ModeString

iSazonov commented:
I wonder why xUnit tests? Pester will be more apropriate and simple for support.

powercode commented:
@SteveL-MSFT 
I found a bug in the `PowerShellPolicyFixture` in test_PSConfiguration.cs, where there is a line
```csharp
Transcription = new Transcription() { 
  EnableInvocationHeader = true, 
  EnableTranscripting = true, 
  OutputDirectory = "c:\tmp" 
},
```

This caused the OutputDirectory to be set to "c:\<tab>mp\", causing intermittent failures, depending on if some other tests had set up transcription correctly before.

I'm fixing this, so that my tests passes :)

powercode commented:
@iSazonov Why would they be easier to support? Having a solution with build/debug/test in an integrated IDE is much more productive, at least for me. I can see the value with Pester when testing end to end scenarios, but the xUnit tests are much nicer when developing or debugging.

iSazonov commented on test/xUnit/csharp/test_PSConfiguration.cs:
```diff
@@ -74,8 +74,8 @@ public PowerShellPolicyFixture()
-                Transcription = new Transcription() { EnableInvocationHeader = true, EnableTranscripting = true, OutputDirectory = "c:\tmp" },
-                UpdatableHelp = new UpdatableHelp() { DefaultSourcePath = "f:\temp" },
+                Transcription = new Transcription() { EnableInvocationHeader = true, EnableTranscripting = true, OutputDirectory = "c:\\tmp" },
+                UpdatableHelp = new UpdatableHelp() { DefaultSourcePath = "f:\\temp" },
```
Perhaps verbatim string literal would be more readable. `@"f:\temp"`

@powercode Many thanks for fixing this!

iSazonov commented:
> Why would they be easier to support?

We have some Pester tests for formatting and the split would add a complicity. I created an Issue for adding xUnit tests but for _public APIs_ to ensure that binary module developers don't get regressions. Formatting is not a contract and we don't need strict control.

powercode committed:
Fixing return type marshalling on createsymlink

iSazonov commented on src/System.Management.Automation/namespaces/FileSystemProvider.cs:
```diff
@@ -7091,6 +7091,7 @@ internal static bool PathIsNetworkPath(string path)
             /// <param name="symbolicLinkFlags">Flag values from SymbolicLinkFlags enum.</param>
             /// <returns>1 on successful creation.</returns>
             [DllImport(PinvokeDllNames.CreateSymbolicLinkDllName, CharSet = CharSet.Unicode, SetLastError = true)]
+            [return: MarshalAs(UnmanagedType.I1)]
```
I wonder why we move from return int to get the marshaling?

powercode replied:
@iSazonov Because int was incorrect. It implies a four byte return, but the signature is 
```C
BOOLEAN CreateSymbolicLinkW(
  LPCWSTR lpSymlinkFileName,
  LPCWSTR lpTargetFileName,
  DWORD   dwFlags
);
```
BOOLEAN is a typedef for BYTE, which is a single byte. marshalling-wise, it is a byte, but logically a bool. And the way to express that in C# native interop, is by using a bool and the 
```CSharp 
[return: MarshalAs(UnmanagedType.I1)]
```
attribute.
This is the kind of bugs that happen to work on certain platforms, and then manifest in weird ways. I think the condition was incorrect to, `result == 1`, since the documentation says 
> If the function succeeds, the return value is nonzero.

But that will work fine with a `bool`. Any non-zero will be true.

iSazonov replied:
I see. Many thanks for fixing this!
Could you contribute more often? :-)

powercode committed:
Only test archive attributes on Windows

powercode committed:
Setting UNIX define on non-windows platforms in test projects.

powercode committed:
Junction test only on windows

SteveL-MSFT reviewed (approved):
LGTM

adityapatwardhan commented on test/xUnit/csharp/test_PSConfiguration.cs:
```diff
@@ -74,8 +74,8 @@ public PowerShellPolicyFixture()
                 ProtectedEventLogging = new ProtectedEventLogging() { EnableProtectedEventLogging = false, EncryptionCertificate = new string[] { "Joe" } },
-                Transcription = new Transcription() { EnableInvocationHeader = true, EnableTranscripting = true, OutputDirectory = "c:\tmp" },
-                UpdatableHelp = new UpdatableHelp() { DefaultSourcePath = "f:\temp" },
+                Transcription = new Transcription() { EnableInvocationHeader = true, EnableTranscripting = true, OutputDirectory = "c:\\tmp" },
```
```suggestion
                Transcription = new Transcription() { EnableInvocationHeader = true, EnableTranscripting = true, OutputDirectory = @"c:\tmp" },
```

adityapatwardhan reviewed (changes requested):
One comment about string literal

adityapatwardhan commented:
@powercode Can you also open an issue in [PowerShell-Docs](https://github.com/MicrosoftDocs/PowerShell-Docs) repo as per the template?

powercode commented:
@adityapatwardhan Docs issue filed, and PR description updated.

powercode commented:
On my end, I'd like more xUnit tests. Much quicker when developing, checked by the compiler, with access to internals, and runs quicker, and often with higher fidelity. 

The possibility do directedly debug a test in the IDE is also really useful.

powercode committed:
Code factor fixes

adityapatwardhan commented:
@powercode I agree that more xUnit tests is generally good. Can these tests be done in Pester?

According to the [testing guidline](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#xunit), we prefer Pester tests over xUnit. xUnit is recommended to be used when expressing the test in Pester is very obtuse and if you are testing an API.

iSazonov commented:
We have #8400 to summarize xUnit gaps. Fill free to feedback there.

powercode commented:
@adityapatwardhan Is the aspect of the dev experience not relevant?

Because this is what make the big difference to me:
![image](https://user-images.githubusercontent.com/3505151/52493820-b4506080-2bcc-11e9-86ab-b7ed5f4938ac.png)

powercode commented:
@TravisEz13 That's the thing - since they tests are written in PowerShell, they exercise a lot of code that is not under test. To me, that is _not_ a good thing. Not to mention refactorings etc.

Oh, well - there I was, all happy that we had XUnit tests...

powercode commented:
So, what happens here?

adityapatwardhan commented:
@powercode We really appreciate the work you have done. This PR is immensely important for PowerShell in regards to performance. 

After discussing this with @PowerShell/powershell-maintainers, we decided we will be contributing to this PR to convert the xUnit tests to Pester and then merge this.

adityapatwardhan committed:
Add FileSystemProvider formatting tests in Pester

adityapatwardhan committed:
Remove xUnit tests for FileSystemProviderFormatting

adityapatwardhan commented:
@powercode I have pushed changes to convert the new xUnit tests to Pester. I also rebased to master to resolve some merge conflicts. Please have a look if I have missed something. And thank you for your patience.

daxian-dbw commented on test/Test.Common.props:
```diff
@@ -10,5 +10,10 @@
+  </PropertyGroup>
+  
+  <PropertyGroup Condition=" '$(IsWindows)' != 'true' ">
+    <DefineConstants>$(DefineConstants);UNIX</DefineConstants>
```
I don't `UNIX` gets used in your changes. Is this added for a general purpose?

powercode replied:
It was for the xUnit tests. They were unwanted.

powercode replied:
But it still makes sense to have it.

adityapatwardhan replied:
@daxian-dbw 
This was used in the xUnit tests. It is not used anymore since we moved the tests to Pester. I am fine with either keeping it or removing it.

powercode commented on src/System.Management.Automation/namespaces/FileSystemProvider.cs:
```diff
@@ -1867,33 +1867,99 @@ private FlagsExpression<FileAttributes> FormatAttributeSwitchParameters()
-                return string.Empty;
+                bool isReparsePoint = InternalSymbolicLinkLinkCodeMethods.IsReparsePoint(fileSystemInfo);
+                bool isHardLink = isReparsePoint || excludeHardLink ? false : InternalSymbolicLinkLinkCodeMethods.IsHardLink(fileSystemInfo);
+                if (!isHardLink)
```
agree

iSazonov replied:
These (ReparsePoint and HardLink) are seems mutually exclusive. The first is less expensive and we could do not evaluate IsHardLink() if we know that it is reparsepoint.

daxian-dbw replied:
The code there has already done that: `isReparsePoint || ...`

powercode commented on src/System.Management.Automation/namespaces/FileSystemProvider.cs:
```diff
@@ -1867,33 +1867,99 @@ private FlagsExpression<FileAttributes> FormatAttributeSwitchParameters()
+        {
+            return instance?.BaseObject is FileSystemInfo fileInfo
+                ? InternalSymbolicLinkLinkCodeMethods.IsReparsePoint(fileInfo)
+                    ? $"{fileInfo.Name} -> {InternalSymbolicLinkLinkCodeMethods.GetTarget(instance).First()}"
```
agree

powercode commented on src/System.Management.Automation/namespaces/FileSystemProvider.cs:
```diff
@@ -1867,33 +1867,99 @@ private FlagsExpression<FileAttributes> FormatAttributeSwitchParameters()
+                bool isDirectory = fileAttributes.HasFlag(FileAttributes.Directory);
+                ReadOnlySpan<char> mode = stackalloc char[]
+                {
+                    isReparsePoint || isHardLink ? 'l' : isDirectory ? 'd' : '-',
```
agree

iSazonov commented on src/System.Management.Automation/namespaces/FileSystemProvider.cs:
```diff
@@ -7909,16 +7976,9 @@ internal static bool IsHardLink(FileSystemInfo fileInfo)
-            }
+            return Platform.IsWindows
+                ? fileInfo.Attributes.HasFlag(System.IO.FileAttributes.ReparsePoint)
+                : Platform.NonWindowsIsSymLink(fileInfo);
```
I wonder why we use `Platform.IsWindows` not `#if UNIX`?

powercode committed:
isHardLink -> isLink

powercode committed:
Changing signature of GetTarget from IEnumerable<string> to string.

The only code path that used multiple return values was !CORECLR.

I removed the obsolete #ifdef-ed code.

powercode commented:
I adressed @daxian-dbw's feedback. 
The other changes LGTM.

adityapatwardhan committed:
Make InternalGetTarget function unix only and remove unreachable code

adityapatwardhan commented:
@daxian-dbw Can you update your review?

daxian-dbw committed:
Minor updates

daxian-dbw commented on src/System.Management.Automation/namespaces/FileSystemProvider.cs:
```diff
@@ -1867,33 +1867,99 @@ private FlagsExpression<FileAttributes> FormatAttributeSwitchParameters()
+                bool isDirectory = fileAttributes.HasFlag(FileAttributes.Directory);
+                ReadOnlySpan<char> mode = stackalloc char[]
+                {
+                    isLink ? 'l' : isDirectory ? 'd' : '-',
```
@powercode I changed it from `(isReparsePoint || isLink) ? 'l' ...` to `isLink ? 'l' ...` because we already have `bool isLink = isReparsePoint || ...` above, so `isReparsePoint` is not needed here again. Please let me know if you have a concern.

powercode replied:
LGTM

powercode replied:
Did that break the tests?

daxian-dbw replied:
> `bool isLink = isReparsePoint || excludeHardLink ? false : InternalSymbolicLinkLinkCodeMethods.IsHardLink(fileSystemInfo);`

The `||` operator has a higher precedence order than `?:`, so we need a parenthesis:
```
bool isLink = isReparsePoint || (excludeHardLink ? false : InternalSymbolicLinkLinkCodeMethods.IsHardLink(fileSystemInfo));
```

daxian-dbw replied:
Pushed a commit to fix that. Let's wait on the CIs.

powercode replied:
Your fix was good, but it exposed a presendense bug where 
```Csharp
isReparsePoint || excludeHardLink ? false : InternalSymbolicLinkLinkCodeMethods.IsHardLink(fileSystemInfo);
```
really is 
```csharp
(isReparsePoint || excludeHardLink) ? false : InternalSymbolicLinkLinkCodeMethods.IsHardLink(fileSystemInfo);
```
and not the intended

```csharp
isReparsePoint || (excludeHardLink ? false : InternalSymbolicLinkLinkCodeMethods.IsHardLink(fileSystemInfo));
```
pushed a fix.

daxian-dbw reviewed (approved):
LGTM

daxian-dbw committed:
Add parenthesis to fix the precedence order

powercode commented:
@daxian-dbw You won the race!
With several seconds :)

adityapatwardhan commented:
@powercode Thanks for your contribution and patience!

lukeb1961 commented:
How will a normal person using Powershell, know that these FT views exist?

iSazonov commented:
@lukeb1961 Please open a feature request (for tab completion) and/or doc isuue.
